samcli/lib/utils/git_repo.py

```diff
@@ -0,0 +1,180 @@
+""" Manage Git repo """
+
+import logging
+import os
+import platform
+import shutil
+import subprocess
+from pathlib import Path
+from typing import Optional
+
+from samcli.lib.utils import osutils
+from samcli.lib.utils.osutils import rmtree_callback
+
+LOG = logging.getLogger(__name__)
+
+
+class CloneRepoException(Exception):
+    """
+    Exception class when clone repo fails.
+    """
+
+
+class CloneRepoUnstableStateException(CloneRepoException):
+    """
+    Exception class when clone repo enters an unstable state.
+    """
+
+
+class GitRepo:
+    """
+    Class for managing a Git repo, currently it has a clone functionality only
+
+    Attributes
+    ----------
+    _url: str
+        The URL of this Git repository, example "https://github.com/aws/aws-sam-cli"
+    _name: str
+        The name of this Git repository, example "aws-sam-cli"
+    _local_path: Path
+        The path of the local clone of this Git repository
+    _clone_attempted: bool
+        whether an attempt to clone this Git repository took place or not
+
+    Methods
+    -------
+    clone(self, clone_dir: Path, clone_name=None, replace_existing=False) -> Path:
+        creates a local clone of this Git repository
+    """
+
+    def __init__(self, url: str, name: str) -> None:
+        self._url: str = url
+        self._name: str = name
+        self._local_path: Optional[Path] = None
+        self._clone_attempted: bool = False
+
+    @property
+    def url(self) -> str:
+        return self._url
+
+    @property
+    def name(self) -> str:
+        return self._name
+
+    @property
+    def local_path(self) -> Optional[Path]:
+        return self._local_path
+
+    @local_path.setter
+    def local_path(self, value: Path) -> None:
+        self._local_path = value
```

Comment:
this getter and setter do not do anything special, we can remove then and make _local_path -> local_path